frontend/catalyst/api_extensions/callbacks.py

```diff
@@ -327,21 +350,36 @@ def total(context, *args, **kwargs):
         jitted_fn = jax.jit(total)
 
     # wraps total which wraps user
-    @functools.wraps(total)
+    @functools.wraps(total, assigned=assigned)
     def back_to_user(*args, **kwargs):
         absextra, absargs, abskwargs = tree_map(shaped_abstractify, (context, args, kwargs))
         try:
             # Find the shape of the return value
             with transient_jax_config({"jax_dynamic_shapes": False}):
-                _, returnshape = jax.make_jaxpr(total, return_shape=True)(
+                _, returnshape = jax.make_jaxpr(jitted_fn, return_shape=True)(
                     absextra, *absargs, **abskwargs
                 )
         except Exception as e:
             name = users_func.__name__
             msg = f"Function {name} must be jax.jit-able."
+            msg += f"But failed with error message {str(e)}."
             raise ValueError(msg) from e
-        annotated = AnnotatedFunction(jitted_fn, returnshape)
-        return jax_jit_callback(annotated, device=dev)(context, *args, **kwargs)
+        annotated = AnnotatedFunctionImpl(jitted_fn, returnshape)
+        with_custom_grad = CallbackWithPotentialCustomGrad(annotated, dev)
+
+        if GradContext.am_inside_grad():
+
+            @with_custom_grad.fwd
+            @accelerate(dev=dev)
+            def vjp_wrapper(context, *args, **kwargs):
+                return jax.vjp(jitted_fn, context, *args, **kwargs)
+
+            @with_custom_grad.bwd
+            @accelerate(dev=dev)
+            def reverse(vjp_func, dy):
+                return vjp_func(dy)
```

Comment:
It probably is, but I'm curious if the context is required:
I'm guessing you want to avoid unnecessary tracing of the function if it's not needed. Is there a chance we could define the attributes and the functions would just be traced on demand?

Reply:
>  Is there a chance we could define the attributes and the functions would just be traced on demand?

Potentially, one of the reasons the context check is required here is because accelerate now recurses. So if we are taking the derivative of function foo, without a check, we would also be taking the derivative of the derivative of foo.

So, a check is needed at some point, whether it is here or inside `with_custom_grad.__call__`. To me, it makes more sense for it to be here. If you want I can dedicate a few cycles to see if I can sink it.

Reply:
> we would also be taking the derivative of the derivative of foo

Is that a bad thing? 🤔  I think that's how jax supports higher level derivates no?

> So, a check is needed at some point, whether it is here or inside with_custom_grad.__call__. To me, it makes more sense for it to be here. If you want I can dedicate a few cycles to see if I can sink it.

That's probably fine, I was just wondering whether we could always assign the fwd/bwd and if they would only be called upon demand.

Reply:
> Is that a bad thing? 🤔 I think that's how jax supports higher level derivates no?

That's infinite derivatives if there is no check.

Reply:
Oh I see, right. You'd have to use jax.jit instead of accelerate to avoid the infinite recursion I suppose. Thanks!